test/relayer.js

```diff
@@ -0,0 +1,92 @@
+    const registry = await deployer.deploy(Registry);
+    relayerModule = await deployer.deploy(TestModule, {}, registry.contractAddress, false, 0);
+    relayerModuleV2 = await deployer.deploy(TestModuleRelayerV2, {}, registry.contractAddress, false, 0);
+    await registry.registerModule(relayerModule.contractAddress, formatBytes32String("RelayerModule"));
+    await registry.registerModule(relayerModuleV2.contractAddress, formatBytes32String("RelayerModuleV2"));
+    wallet = await deployer.deploy(Wallet);
+    await wallet.init(owner.address, [relayerModule.contractAddress, relayerModuleV2.contractAddress]);
+  });
+
+  describe("RelayerModule", () => {
+    it("should fail to relay when target of _data != _wallet", async () => {
+      const params = [await getRandomAddress(), 4]; // the first argument is not the wallet address, which should make the relaying revert
+      await assert.revertWith(
+        manager.relay(relayerModule, "setIntOwnerOnly", params, wallet, [owner]), TARGET_OF_DATA_NOT_WALLET_REVERT_MSG,
+      );
+    });
+
+    it("should fail to relay when _data is less than 36 bytes", async () => {
+      const params = []; // the first argument is not the wallet address, which should make the relaying rever
+      await assert.revertWith(
+        manager.relay(relayerModule, "clearInt", params, wallet, [owner]), INVALID_DATA_REVERT_MSG,
+      );
+    });
+
+    it("should fail to relay a duplicate transaction", async () => {
+      const params = [wallet.contractAddress, 2];
+      const nonce = await getNonceForRelay();
+      const relayParams = [relayerModule, "setIntOwnerOnly", params, wallet, [owner],
+        global.accounts[9].signer, false, 2000000, nonce];
+      await manager.relay(...relayParams);
+      await assert.revertWith(
+        manager.relay(...relayParams), DUPLICATE_REQUEST_REVERT_MSG,
+      );
+    });
+  });
+
+  describe("RelayerModuleV2", () => {
+    it("should fail to relay when target of _data != _wallet", async () => {
+      const params = [await getRandomAddress(), 4]; // the first argument is not the wallet address, which should make the relaying revert
+      await assert.revertWith(
+        manager.relay(relayerModuleV2, "setIntOwnerOnly", params, wallet, [owner]), TARGET_OF_DATA_NOT_WALLET_REVERT_MSG_V2,
+      );
+    });
+
+    it("should fail to relay when _data is less than 36 bytes", async () => {
+      const params = []; // the first argument is not the wallet address, which should make the relaying rever
+      await assert.revertWith(
+        manager.relay(relayerModuleV2, "clearInt", params, wallet, [owner]), INVALID_DATA_REVERT_MSG,
+      );
+    });
+
+    it("should fail to relay a duplicate transaction", async () => {
+      const params = [wallet.contractAddress, 2];
+      const nonce = await getNonceForRelay();
+      const relayParams = [relayerModule, "setIntOwnerOnly", params, wallet, [owner],
+        global.accounts[9].signer, false, 2000000, nonce];
```

Comment:
Why pass a nonce that will be generated in `manager.relay` the same way?

Reply:
We want the two calls to `manager.relay()` to be using the same nonce. If you let `manager.relay()` compute its own nonce during the second call, the computed nonce will be different than the one used in the first call.